src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Boxing/BoxingInstructions.cs

```diff
@@ -60,74 +57,102 @@ public static ParameterBoxingInstructions[] GetBoxingInstructions(MethodInfo met
                     // To enable it in the future add a new special case token and when rewriting IL
                     // emit a ldobj instruction for it.
                     //
-                    instructions.Add(SpecialCaseBoxingTypes.Unknown);
+                    thisBoxingInstructions = SpecialCaseBoxingTypes.Unknown;
                 }
                 else
                 {
-                    methodParameterTypes.Insert(0, thisType);
-                    // Implicit this isn't a formal parameter, so offset by one.
-                    formalParameterPosition = -1;
+                    thisBoxingInstructions = GetBoxingInstructionsFromReflection(method, thisType, out _);
                 }
+
+                instructions[index++] = thisBoxingInstructions;
             }
 
-            foreach (Type paramType in methodParameterTypes)
+            //
+            // A signature decoder will be used to determine boxing tokens for formal parameter types that cannot be determined from standard
+            // reflection alone. The boxing tokens generated from this decoder should only be used to fill in these gaps
+            // as it is not a comprehensive decoder and will produce an unsupported boxing instruction for any types not explicitly mentioned
+            // in BoxingTokensSignatureProvider's summary.
+            // 
+            Lazy<ParameterBoxingInstructions[]?> signatureDecoderInstructions = new(() =>
             {
-                if (paramType.IsByRef ||
-                    paramType.IsByRefLike ||
-                    paramType.IsPointer)
-                {
-                    instructions.Add(SpecialCaseBoxingTypes.Unknown);
-                }
-                else if (paramType.IsGenericParameter)
+                ParameterBoxingInstructions[]? instructions = GetAncillaryBoxingInstructionsFromMethodSignature(method);
+                if (instructions != null &&
+                    instructions.Length != formalParameters.Length)
                 {
-                    instructions.Add(SpecialCaseBoxingTypes.Unknown);
+                    Debug.Fail("Signature decoder results have an unexpected number of formal parameters");
```

Comment:
If `instructions` is null, do we also want this debug line? As it currently stands, it'll return null either way, but the message would be omitted

Reply:
`GetAncillaryBoxingInstructionsFromMethodSignature` returning null is an expected code path so it doesn't indicate a problem (whereas returning a mismatch on the number of parameters should *never* happen but we still need to sanity check it).

There is one code path inside `GetAncillaryBoxingInstructionsFromMethodSignature` where returning null is unexpected, so I'll go ahead an add a debug message over there